dsl/flatten.cc

```diff
@@ -0,0 +1,322 @@
+            auto &methods = curMethodSet();
+            int staticLevel = 0;
+            if (isMethodModifier(*send) && send->args.size() >= 1) {
+                // if this is a method modifier like `private` or `protected`, then we don't need to add a new scope
+                // when we traverse the method itself, so add it to the ignore set...
+                skipMethods.insert(send->args[0].get());
+                auto methodDef = ast::cast_tree<ast::MethodDef>(send->args.front().get());
+                ENFORCE(methodDef != nullptr);
+                staticLevel = computeStaticLevel(methodDef);
+            }
+
+            // we should only move methods that are nested, so if the method stack is empty, then don't bother adding
+            // space in the move queue and record the index as -1
+            if (methods.stack.size() == 0) {
+                methods.stack.emplace_back(-1, staticLevel);
+            } else {
+                methods.stack.emplace_back(methods.methods.size(), staticLevel);
+                methods.methods.emplace_back();
+            }
+        }
+        return send;
+    }
+
+    unique_ptr<ast::Expression> postTransformSend(core::Context ctx, unique_ptr<ast::Send> send) {
+        if (send->fun == core::Names::sig() || isMethodModifier(*send)) {
+            auto &methods = curMethodSet();
+            ENFORCE(!methods.stack.empty());
+
+            // we may not need to move this method at all: check first
+            if (methods.stack.back().idx == -1) {
+                methods.stack.pop_back();
+                return send;
+            }
+
+            // otherwise, we have a spot in the queue for it that has not yet been filled in
+            ENFORCE(methods.methods.size() > methods.stack.back().idx);
+            ENFORCE(methods.methods[methods.stack.back().idx] == nullptr);
+
+            if (isMethodModifier(*send) && send->args.size() >= 1) {
+                auto methodDef = ast::cast_tree<ast::MethodDef>(send->args.front().get());
+                ENFORCE(methodDef);
+                int level = methods.stack.back().staticLevel;
+                methodDef->setIsSelf(level > 0);
+                auto loc = methodDef->loc;
+                auto declLoc = methodDef->declLoc;
+                while (level > 1) {
+                    ast::ClassDef::RHS_store rhs;
+                    rhs.emplace_back(move(send->args.front()));
+                    auto nestedClass =
+                        ast::MK::Class(loc, declLoc,
+                                       make_unique<ast::UnresolvedIdent>(core::Loc::none(), ast::UnresolvedIdent::Class,
+                                                                         core::Names::singleton()),
+                                       {}, std::move(rhs), ast::ClassDefKind::Class);
+                    send->args.front() = move(nestedClass);
+                    level--;
+                }
```

Comment:
This logic is temporary and is used for replicating the current (non-correct) behavior: this loop is used to wrap a `class << self; ...; end` around methods. This logic (which is kind of hairy!) will become a lot simpler when we implement the correct semantics.